### DISCUSSION TITLE
Is there a way to avoid 'conceal' when using ItemsService within nodejs?

### DISCUSSION THREAD

**Discussion by pzehle** [directus/directus#17632] [Archive]
### Checklist

- [X] I'm using the [latest version of Directus](https://github.com/directus/directus/releases)
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes the problem.
- [X] I've completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).

### Describe the Bug

I have my entire environment of email management and templates with PostMark. Directus is definitely not made for the connection to this platform, as the simple action of sending a password recovery token assumes you will send the email from Directus itself. This is not my case. I managed to recreate the entire flow Directus uses to generate the tokens and validate them, the problem is whenever I want to create the 'hash' to include in the JWT that I will send in the email, the password I get for the user is just '***********' concealed. Is there a way to avoid this conceal call using the items service? Or should I just connect directly to the database just to retrieve the already-hashed password?

Thank you.

### To Reproduce

Get the password of a user using ItemsService

### Hosting Strategy

Self-Hosted (Custom)

**Comment by pzehle** [comment]
Whether if one can pass `conceal: false` as an option in the items service, have some kind of 'util' to get the hash of the password, or even use additional options in the `.env` file, such as `EMAIL_ADDITIONAL_HEADERS`, because using Postmark you can pass headers when you use SMTP, to decide which template, stream, address, etc you will be using when sending an email. Either of these options would be nice, but to be honest I am kind of surprised that it is impossible to get the user's password from within the code. I understand it is a security matter, but it should be the decision of the developer whether to use this or not. If you take React for example, you can insert HTML into components, and since React considers this as an insecure practice, the name of the property becomes `dangerouslySetInnerHTML`, telling you that you are using an insecure property, but you are still able to use it because in some cases you need to. I feel like getting the password of the user, which in any case you can't and won't ever get the plaintext password, should be something possible within Directus' environment.

**Reply by pzehle** [reply]
I am not entirely sure this is not an issue, since one might expect to get raw information from the database using the Items Service, and Directus, by default, concealing information should not, in my opinion (and I can't be the only one), be something you just "have to live with"...

**Comment by br41nslug** [comment]
There is no way to avoid the conceal using the provided services or API. This property is applied to any field which would be considered a security vulnerability when disclosed over the API or mail like passwords or access tokens.

While it should not be necessary in most cases, you can get the raw values by using the `knex` instance to directly fetch the field from the database.

for example in a hook you *could* do something like:
```js
export default (_actions, { database }) => {
    // ...
    const data = await database
        .select('password')
        .from('directus_users')
        .where('email', '=', 'admin@example.com');
    // ...
}
```

**Reply by pzehle** [reply]
Thanks. I figured something like this out. Still I believe, that if you are able to do this within the code, you _should_ be able to do it with the normal services anyway.

Thanks for your time and explanation.

**Reply by br41nslug** [reply]
It might be worth a feature request to see if the community agrees 😄

**Reply by pzehle** [reply]
It is not really about the password only. It is about the "blocking" of a field and no "built-in" way to get it if not connecting directly to the database, which in that case, breaks the meaning of Directus, at least in my opinion.

The only reason I need the password for is, to send the JWT for password recovery using Postmark. Since Directus assumes you will use their email system, you are not able to send API requests to other email providers; and even using SMTP, at least with Postmark, you are not able to add additional headers to select a template or other stuff.
So as I said, it is not only the password: Is the blocking of a field just because you decided it is better that way, and the inability to use another email provider via API.

Anyway, this is solved, and again, thanks :)
